### PR TITLE
Feature conditions all widgets

### PR DESCRIPTION
**Pull request by AdityaKhatri**
Add conditions and condition checks for all widgets except geo.
Should first merge https://github.com/the-deep/client/pull/175.

### CODE REVIEW

File: src/widgets/conditionalWidget/date.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        type: 'date',
+        title: 'Before date',
+    }],
+    test: ({ value }, { value: attrValue }) => (
+        compareDate(value, attrValue) < 0
+    ),
+};
+
+const isInBetween = {
+    title: 'In in between',
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/date.js]: WOW English

File: src/widgets/conditionalWidget/matrix1d.js [hunk 2]
```diff
@@ -1,13 +1,16 @@
...
         title: r.title,
     }))
 );
 
-const getSubpillarOptions = widgetData => (
-    widgetData.rows.reduce((acc, r) => [
-        ...r.cells.map(c => ({
+const getSubpillarOptions = ({ rows = emptyArray } = {}) => (
+    rows.reduce((acc, r) => [
+        ...((r || emptyObject).cells || emptyArray).map(c => ({
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix1d.js]: r should have a default value instead
Avoid using variable names without meaning like 'r'

File: src/widgets/conditionalWidget/matrix1d.js [hunk 3]
```diff
@@ -41,6 +44,14 @@ const containsSubpillar = {
         keySelector: d => d.key,
         labelSelector: d => d.title,
     }],
+    test: ({ value = {} }, { subpillar }) => (
+        Object.keys(value).some((p) => {
+            const subpillars = value[p] || emptyObject;
+            return Object.keys(subpillars).some(sp => (
+                sp === subpillar ? subpillars[sp] : false
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix1d.js]: very lazy variable naming

File: src/widgets/conditionalWidget/matrix2d.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+        type: 'select',
+        title: 'Subdimension',
+        options: getSubdimensionOptions,
+        keySelector: d => d.key,
+        labelSelector: d => d.title,
+    }],
+    test: ({ value = {} }, { subdimension }) => {
+        const dimensionOfSubdimension = Object.keys(value)
+            .find(v => (value[v] || emptyObject)[subdimension] !== undefined);
+        if (!dimensionOfSubdimension) {
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix2d.js:78]: It's better to check the length of array

File: src/widgets/conditionalWidget/matrix2d.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        type: 'select',
+        title: 'Sector',
+        options: getSectorOptions,
+        keySelector: d => d.key,
+        labelSelector: d => d.title,
+    }],
+    test: ({ value = {} }, { sector }) => (
+        Object.keys(value).some((v) => {
+            const subdimen = value[v];
+            if (isObjectEmpty(subdimen)) {
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix2d.js:98]: Returning true/false value can be avoided (as said by Dahal)

**Review by tnagorra — CHANGES_REQUESTED** [review]

File: src/widgets/conditionalWidget/matrix2d.js [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        type: 'select',
+        title: 'Subsector',
+        options: getSubsectorOptions,
+        keySelector: d => d.key,
+        labelSelector: d => d.title,
+    }],
+    test: ({ value = {} }, { subsector }) => (
+        Object.keys(value).some((v) => {
+            const subdimen = value[v];
+            if (isObjectEmpty(subdimen)) {
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix2d.js:122]: Same as previous comment

File: src/widgets/conditionalWidget/matrix2d.js [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+                return false;
+            }
+
+            return Object.keys(subdimen).some((sd) => {
+                const subdimenSector = subdimen[sd];
+                if (isObjectEmpty(subdimenSector)) {
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/matrix2d.js:128]: Same as previous comment

File: src/widgets/conditionalWidget/scale.js [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            key: 'upperScale',
+            type: 'select',
+            title: 'Upper Scale',
+            options: getScaleOptions,
+            keySelector: d => d.key,
+            labelSelector: d => d.label,
+        },
+    ],
+    test: ({ value }, { lowerScale, upperScale }, { scaleUnits = [] }) => (
+        (scaleUnits.findIndex(s => s.key === value) >=
```
Comment (tnagorra) [inline src/widgets/conditionalWidget/scale.js]: very lazy thing to do

File: src/widgets/conditionalWidget/matrix2d.js (hunk 5 above)
Comment (AdityaKhatri) [reply src/widgets/conditionalWidget/matrix2d.js:98]: Dahal suggests the code is cleaner this way.

File: src/widgets/conditionalWidget/matrix2d.js (hunk 6 above)
Comment (AdityaKhatri) [reply src/widgets/conditionalWidget/matrix2d.js:122]: Same as previous comment

File: src/widgets/conditionalWidget/matrix2d.js (hunk 7 above)
Comment (AdityaKhatri) [reply src/widgets/conditionalWidget/matrix2d.js:128]: Same as previous comment

**Review by tnagorra — APPROVED** [review]

**Pull request merged**